---
title: 
output: html_document
---
# <span style="color: darkseagreen"> Colocalization Analysis on Significant Genes</span>
## <span style="color: darkolivegreen"> Run coloc for significant genes in S-PrediXcan for METS Sleep Project </span>

## <span style="color: darkolivegreen"> 1. Retrieve cis-QTL data from GTEx Portal </span>
* [GTEx Portal](https://gtexportal.org/home/datasets#filesetFilesDiv14)

## <span style="color: darkolivegreen"> 2. Convert eQTL file to correct format </span>
* <span style="color: darkslategray"> only include snps associated with the significant gene </span>
* <span style="color: darkslategray"> remove insertion/deletion snps </span>
* <span style="color: darkslategray"> Information about the data in the eQTL file can be found on the GTEx website in the Single Tissue cis-QTL Data README file  </span>

```{r,eval= FALSE}
library(data.table)
library(dplyr)
library(coloc)
library(hash)

"%&%" = function(a,b) paste(a,b,sep="")


#read in the cis eQTL file
path <- '/home/grace/coloc/eQTL_data/Brain_Hypothalamus.allpairs.txt.gz'
data_eQTL <- fread(path)

#adding necessary columns
data_eQTL <- data_eQTL %>%
  mutate(`#CHROM` = sapply(strsplit(variant_id,"_"), `[`, 1), #chromosome number
         POS = sapply(strsplit(variant_id,"_"), `[`, 2), #position
         REF = sapply(strsplit(variant_id,"_"), `[`, 3), #reference allele, w/ indels
         A1 = sapply(strsplit(variant_id,"_"), `[`, 4)) %>% #alternate allele w/ indels
  rename(OBS_CT = ma_count, #number of allele observations
         A1_FREQ = maf, #A1 allele frequency
         BETA = slope, #regression coefficient for A1 allele
         SE = slope_se, #standard error
         P = pval_nominal) #p-value

data_eQTL <- data_eQTL %>%
  filter(nchar(A1)==1, #remove indels from alt alleles
         nchar(REF) == 1) %>% #remove indels from ref alleles
  mutate(chr = substring(`#CHROM`, 4), #create ID column
         ALT = A1) #create ALT column
data_eQTL <- data_eQTL %>%
  mutate(ID = (chr %&% "_" %&% POS))

 #creating new eQTL file with correct data
eQTLpath <- '/home/grace/coloc/eQTL_data/Brain_Hypo_eQTL.csv'
fwrite(data_eQTL, eQTLpath, append=FALSE, quote = 'auto', sep=',') 
```

## <span style="color: darkolivegreen"> 3. Convert GWAS file to correct format </span>
* <span style="color: darkslategray"> only include snps associated with the significant gene</span>
* <span style="color: darkslategray"> remove insertion/deletion snps</span>

```{r, eval =FALSE}
library(data.table) 
library(dplyr)
library(coloc)
library(hash)

"%&%" = function(a,b) paste(a,b,sep="")

#reformatting the GWAS data
data_GWAS <- fread("/home/grace/PrediXcan/hg38_PUKBB_META_chronotype.tsv.gz") #GWAS file, build 38

data_GWAS <- data_GWAS %>%
  mutate(CHR = substring(chromosome,4), #get chromosome number
         BP = sapply(strsplit(panel_variant_id,"_"), `[`, 2)) %>% #get position 
  filter(nchar(effect_allele) == 1, #filter out indels
         nchar(non_effect_allele) == 1) %>% #filter out indels
  rename(FREQ = frequency, #rename to correct column names
         BETA = effect_size,
         SE = standard_error,
         P = pvalue)

data_GWAS <- data_GWAS %>%
  mutate(ID =(CHR %&% "_" %&% BP), 
         SNP_hg38 = BP) %>% #makes ID column
  rename(A1 = effect_allele, #rename to correct column names
         A2 = non_effect_allele)

GWASpath <- "/home/grace/coloc/gwasFiles/META_chronotype_coloc"
fwrite(data_GWAS, GWASpath, append=FALSE, quote ='auto', sep=',') #write new GWAS file
```

## <span style="color: darkolivegreen"> 4. Filter eQTL data to SNPs related to significant gene only </span>

```{r,eval=FALSE}
library(data.table)
library(dplyr)
library(coloc)
library(hash)

"%&%" = function(a,b) paste(a,b,sep="")

eQTL_Bigdir = "/home/grace/coloc/eQTL_data/Brain_Hypo_eQTL.csv" 
eQTLs <- fread(eQTL_Bigdir) #read in eQTLs from Hypothalamus 

eQTLs <- eQTLs %>%
  filter(gene_id == "ENSG00000164924.17") %>% #only use snps with ensembl id of the significant gene
  arrange(P, by_group = FALSE) #arrange by p value
fwrite(eQTLs, "/home/grace/coloc/eQTL_data/YWHAZ_data.csv", append=FALSE,quote = "auto", sep=",") #write new eQTL file
```
## <span style="color: darkolivegreen"> 5. Retrieve summary stat data from gene and phenotype of interest</span>

```{r}
library(data.table)
library(dplyr)
library(coloc)
library(hash)

"%&%" = function(a,b) paste(a,b,sep="")

GWASdir = "/home/grace/coloc/gwasFiles/META_chronotype_coloc"
eQTLdir = "/home/grace/coloc/eQTL_data/YWHAZ_data.csv"

eQTLs = fread(eQTLdir) #read in significant gene eQTL
gwas = fread(GWASdir)

#get intersecting SNPs
joined = inner_join(eQTLs,gwas, by=c("ID"="ID"))

joined <- arrange(joined, P.y, by_group=FALSE)


#check for complement bases
#build hash table (like a Python dictionary)

bases = hash()
bases[["A"]] <- "T"
bases[["C"]] <- "G"
bases[["G"]] <- "C"
bases[["T"]] <- "A"



#pull SNPs that  match (assumes C|G and A|T SNPs aren't flipped)
match = joined[(joined$A1.x==joined$A1.y & joined$REF == joined$A2),]

#remove ambiguous strand SNPs (A/T or C/G)
a = joined[!(joined$REF=="A" & joined$ALT=="T") & !(joined$REF=="T" & joined$ALT=="A") & !(joined$REF=="C" & joined$ALT=="G") &
             !(joined$REF=="G" & joined$ALT=="C") ]

#of non-ambiguous, pull SNPs that are flipped (or the complement bases match or are flipped)

compmatch = a[(a$A1.x==values(bases,keys=a$A1.y) & a$REF == values(bases,keys=a$A2)),]

flipped = a[(a$A1.x==a$A2 & a$REF == a$A1.y),]
compflipped = a[(a$A1.x==values(bases,keys=a$A2) & a$REF == values(bases,keys=a$A1.y)),]

#if flipped, change sign of cancer beta, check to see if any SNPs are in flipped df's with if stmt
if(dim(flipped)[1] > 0){
    flipped = mutate(flipped,BETA.y = -1*BETA.y)
}
if(dim(compflipped)[1] > 0){
     compflipped = mutate(compflipped,BETA.y = -1*BETA.y)
}

#bind all and sort by position
matchsnps = rbind(match, compmatch, flipped, compflipped) %>% arrange(POS)

#matchsnps <- matchsnps %>%
  #na.omit()

matchsnps <- matchsnps %>%
  distinct(BP, .keep_all = TRUE) %>%
  select(-n_cases)
matchsnps <- matchsnps %>%
  na.omit()
```

## <span style="color: darkolivegreen"> 6. Format each set of summary stats for coloc </span>
* <span style="color: darkslategray"> Run `vignette("ao2_data", package="coloc")` for more details </span>

```{r}
gwascoloc = list("beta" = matchsnps$BETA.y, "varbeta" = (matchsnps$SE.y)^2, "snp" = matchsnps$ID, "position" = matchsnps$BP,
                 "type" = "cc")
eqtlcoloc = list("beta" = matchsnps$BETA.x, "varbeta" = (matchsnps$SE.x)^2, "snp" = matchsnps$ID, "position" = matchsnps$POS,
                 "type" = "quant", "N" = matchsnps$OBS_CT[1], "MAF" = matchsnps$A1_FREQ, "sdY"=1 )


plot_dataset(gwascoloc)
plot_dataset(eqtlcoloc)
```

## <span style="color: darkolivegreen"> 7. Run coloc assuming a single causal variant </span>

```{r}
my.res = coloc.abf(dataset1=gwascoloc, dataset2=eqtlcoloc)
my.res
sensitivity(my.res,"H4 > 0.9")
```

## <span style="color: darkolivegreen"> 8. Add LD info to data frames for coloc.susie (multiple causal SNPs) </span>

```{r}
refdir = "/home/grace/coloc/missingness_hwe_steps/"
#make a list of EUR individuals in 1000 Genomes
pops = fread(refdir %&% "02geno_0.001_maf_0.05_filtered.fam")
eurlist = mutate(pops, V1=0) %>% dplyr::select(V1, V2)
fwrite(eurlist,"/home/grace/coloc/EUR_list",col.names=FALSE, sep="\t")

#need chromosome code, start of range, end of range, set id for plink call
chr <- c(8)
first <- c(100916523)
last <- c(100953388)
id <- c('YWHAZ')
range <- data.frame(chr, first, last, id)
fwrite(range, "/home/grace/coloc/range.txt", col.names = FALSE, sep='\t')

#system call to plink to retrieve desired SNPs and individuals in .raw format
#genotypes must vary to be useful, maf>0.01

system("plink --bfile " %&% refdir %&% "02geno_0.001_maf_0.05_filtered --extract range /home/grace/coloc/range.txt --keep /home/grace/coloc/EUR_list --maf 0.01 --recode A --make-just-bim --out /home/grace/coloc/EUR_coloc-region --allow-extra-chr")

#read the EUR .raw file generated by plink into R.
geno = fread("/home/grace/coloc/EUR_coloc-region.raw")

#we want to put genotypes in NxP matrix (N=people, P=snps) with no other columns
genomat = as.matrix(geno[,7:length(geno)])
#now filter the gwascoloc and eqtlcoloc df's to those SNPs that were in EUR (in genomat)
#get a list of SNPs in genomat
snplist = colnames(genomat)
#remove the last 2 characters (_N) from genomat colnames to match topsleep with substr()
snplist = substr(snplist, 1, nchar(snplist)-2)
#rename col names of R to match coloc df's
colnames(genomat) = snplist

#check eQTL and 1000G EUR alleles (ask, what is coded as 1?), if need to, flip BETA signs.
#col 5 in bim is A1 (assigned 1 in dosage raw file)
bim = fread("/home/grace/coloc/EUR_coloc-region.bim")

bim <- bim %>%
  mutate(ID = V1 %&% "_" %&% V4)

susiesnpsbim = inner_join(matchsnps, bim, by=c("ID"="ID"))

#pull SNPs that  match (assumes C|G and A|T SNPs aren't flipped)
matchbim = susiesnpsbim[(susiesnpsbim$A1.x==susiesnpsbim$V5 & susiesnpsbim$REF == susiesnpsbim$V6),]

#remove ambiguous strand SNPs (A/T or C/G)
b = susiesnpsbim[!(susiesnpsbim$REF=="A" & susiesnpsbim$ALT=="T") & !(susiesnpsbim$REF=="T" & susiesnpsbim$ALT=="A") & !(susiesnpsbim$REF=="C" & susiesnpsbim$ALT=="G") & !(susiesnpsbim$REF=="G" & susiesnpsbim$ALT=="C") ]

#of non-ambiguous, pull SNPs that are flipped (or the complement bases match or are flipped)
compmatchbim = b[(b$A1.x==values(bases,keys=b$V5) & b$REF == values(bases,keys=b$V6)),]
flippedbim = b[(b$A1.x==b$V6 & b$REF == b$V5),]
compflippedbim = b[(b$A1.x==values(bases,keys=b$V6) & b$REF == values(bases,keys=b$V5)),]

#if flipped, change sign of cancer and eqtl beta, check to see if any SNPs are in flipped df's with if stmt
if(dim(flippedbim)[1] > 0){
    flippedbim = mutate(flippedbim,BETA.x = -1*BETA.x,BETA.y = -1*BETA.y)
}
if(dim(compflippedbim)[1] > 0){
     compflippedbim = mutate(compflippedbim,BETA.x = -1*BETA.x,BETA.y = -1*BETA.y)
}

#bind all and sort by position
matchbimsnps = rbind(matchbim, compmatchbim, flippedbim, compflippedbim) %>% arrange(POS)

#update snplist
snplist = matchbimsnps$V2
#filter genomat to snps in snplist
x = genomat[,colnames(genomat) %in% snplist]
#calculate the correlation matrix of the genotypes, this is needed for susie
R = cor(x)

#filter the matchbimsnps df to just the snps in snplist with dplyr
susiesnps = filter(matchbimsnps, V2 %in% snplist)
#add LD to filtered coloc df's
gwascolocsusie = list("beta" = susiesnps$BETA.y, "varbeta" = (susiesnps$SE.y)^2, "snp" = susiesnps$V2, "position" = susiesnps$POS,
                 "type" = "cc", "LD"=R, "N" = 100000) #need to add N, check ref later
eqtlcolocsusie = list("beta" = setNames(susiesnps$BETA.x, susiesnps$V2), "varbeta" = setNames(susiesnps$SE.x^2, susiesnps$V2), "snp" = susiesnps$V2,
                      "position" = susiesnps$POS,"type" = "quant", "N" = susiesnps$OBS_CT[1], "MAF" = susiesnps$A1_FREQ, "sdY"=1, "LD"= R)

check_dataset(gwascolocsusie,req="LD")

check_dataset(eqtlcolocsusie,req="LD")

plot_dataset(gwascolocsusie)

plot_dataset(eqtlcolocsusie)
```

## <span style="color: darkolivegreen"> 9. Run susie on each df and then run coloc </span>

```{r}
#run susie
sgwas = runsusie(gwascolocsusie)
seqtl = runsusie(eqtlcolocsusie)
#up to 10 credible sets by default
summary(sgwas)
summary(seqtl)
#run coloc
susie.res=coloc.susie(sgwas, seqtl)
print(susie.res$summary)
if(!is.na(susie.res)[1]){
  #retrive row numbers with H4>0.5
  sigrows = which(susie.res$summary$PP.H4.abf > 0.5)
  sigrows
  
  #sensitivity plot for row 1
  sensitivity(susie.res,"H4 > 0.5",row=1,dataset1=gwascolocsusie,dataset2=eqtlcolocsusie)
  
  #sensitivity plots for rows with H4>0.5
  for(i in sigrows){
    sensitivity(susie.res,"H4 > 0.5",row=i,dataset1=gwascolocsusie,dataset2=eqtlcolocsusie)
  }
  }
```

## <span style="color: darkolivegreen"> 10. Run susie with lower coverage on each df and then run coloc </span>

```{r}
#run susie
sgwas = runsusie(gwascolocsusie, coverage=0.1)
seqtl = runsusie(eqtlcolocsusie, coverage=0.1)
#up to 10 credible sets by default
summary(sgwas)
summary(seqtl)
#run coloc
susie.res=coloc.susie(sgwas, seqtl)
print(susie.res$summary)
if(!is.na(susie.res)[1]){
  #retrieve row numbers with H4>0.4
  sigrows = which(susie.res$summary$PP.H4.abf > 0.4)
  sigrows
  
  #sensitivity plot for row 1
  sensitivity(susie.res,"H4 > 0.5",row=1,dataset1=gwascolocsusie,dataset2=eqtlcolocsusie)
  
  #sensitivity plots for rows with H4>0.4
  for(i in sigrows){
    sensitivity(susie.res,"H4 > 0.4",row=i,dataset1=gwascolocsusie,dataset2=eqtlcolocsusie)
  }
}

```
